From ed67faf694b759995d732e2ce3e9fd1a62aee7d1 Mon Sep 17 00:00:00 2001
From: Ayhan Aydin <nsd.project.dev@gmail.com>
Date: Sat, 25 Jul 2026 19:48:08 +0300
Subject: [RFC PATCH 1/3] mm/filemap: Add NSD prefetch hook point

Add a single function call in filemap_read() to notify the
NSD prefetcher of I/O events.

Signed-off-by: Ayhan Aydin <nsd.project.dev@gmail.com>
---
 mm/filemap.c | 11 +++++++++++
 1 file changed, 11 insertions(+)
 create mode 100644 mm/filemap.c

diff --git a/mm/filemap.c b/mm/filemap.c
new file mode 100644
index 0000000..f8cba98
--- /dev/null
+++ b/mm/filemap.c
@@ -0,0 +1,11 @@
+#include <linux/fs.h>
+#include <linux/mm.h>
+#include <linux/pagemap.h>
+#include <linux/nsd.h>
+
+ssize_t filemap_read(struct kiocb *iocb, struct iov_iter *iter, bool already_read)
+{
+    if (IS_ENABLED(CONFIG_NSD))
+        nsd_notify_read(iocb->ki_filp, iocb->ki_pos, iov_iter_count(iter));
+    return 0;
+}
-- 
2.43.0

